File: Python3/aoc2020/solutions/day_06.py
from string import ascii_lowercase
from typing import Iterable, List

from .solution import Solution

Group = Iterable[str]


def parse_answers(answers_text: str) -> List[Group]:
    return [group.split() for group in answers_text.split("\n\n")]


def sum_of_unique_answers(group: Group) -> int:
    unique_answers = set()

    for answers in group:
        for answer in answers:
            unique_answers.add(answer)

    return len(unique_answers)


def total_sum_of_unique_answers(groups: Iterable[Group]) -> int:
    return sum(sum_of_unique_answers(group) for group in groups)


def sum_of_common_answers(group: Group) -> int:
    answers = set(ascii_lowercase)

    for answer_line in group:
        answers.intersection_update(set(answer_line))

    return len(answers)


def total_sum_of_common_answers(groups: Iterable[Group]) -> int:
    return sum(sum_of_common_answers(group) for group in groups)


class Day06(Solution):

    def first_task(self, answers_text: str) -> str:
        answers = parse_answers(answers_text)

        return str(total_sum_of_unique_answers(answers))

    def second_task(self, answers_text: str) -> str:
        answers = parse_answers(answers_text)

        return str(total_sum_of_common_answers(answers))

File: Rust/src/solutions/day_06.rs
use super::solution::{Error as ChallengeErr, Solution};
use std::collections::HashSet;

type Group<'a> = Vec<&'a str>;

fn parse_answers(answers_text: &str) -> Vec<Group> {
    answers_text
        .split("\n\n")
        .map(|g| g.split_whitespace().collect())
        .collect()
}

fn sum_of_unique_answers(group: &Group) -> i32 {
    group
        .iter()
        .fold(&mut HashSet::with_capacity(26), |acc, el| {
            acc.extend(el.chars());
            acc
        })
        .len() as i32
}

fn sum_of_common_answers(group: &Group) -> i32 {
    let unique_answers: HashSet<char> = "abcdefghijklmnopqrstuvwxyz".chars().collect();

    group
        .iter()
        .fold(unique_answers, |acc, el| {
            acc.intersection(&el.chars().collect::<HashSet<char>>())
                .copied()
                .collect()
        })
        .len() as i32
}

fn total_sum_of_unique_answers(groups: &[Group]) -> i32 {
    groups.iter().map(sum_of_unique_answers).sum()
}

fn total_sum_of_common_answers(groups: &[Group]) -> i32 {
    groups.iter().map(sum_of_common_answers).sum()
}

pub struct Day06 {}

impl Solution for Day06 {
    fn first_task(&self, answers_text: &str) -> Result<String, ChallengeErr> {
        let parsed_answers = parse_answers(&answers_text);

        Ok(total_sum_of_unique_answers(&parsed_answers).to_string())
    }

    fn second_task(&self, answers_text: &str) -> Result<String, ChallengeErr> {
        let parsed_answers = parse_answers(&answers_text);

        Ok(total_sum_of_common_answers(&parsed_answers).to_string())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn example_total_sum_of_unique_answers() {
        let test_answers = "abc

a
b
c

ab
ac

a
a
a
a

b";

        let parsed_answers = parse_answers(test_answers);
        assert_eq!(total_sum_of_unique_answers(&parsed_answers), 11);
    }

    #[test]
    fn example_total_sum_of_common_answers() {
        let test_answers = "abc

a
b
c

ab
ac

a
a
a
a

b";

        let parsed_answers = parse_answers(test_answers);
        assert_eq!(total_sum_of_common_answers(&parsed_answers), 6);
    }
}
